blockchain/vm/contracts.go

```diff
@@ -448,51 +434,29 @@ func (c *feePayer) GetRequiredGasAndComputationCost(input []byte) (uint64, uint6
 	return params.FeePayerGas, params.FeePayerComputationCost
 }
 
-func (c *feePayer) Run(input []byte) ([]byte, error) {
-	// Run function should not be called. Instead of this function, RunFeePayerContract should be called.
-	logger.Error("should not be reached here")
-	return nil, nil
-}
-
-func RunFeePayerContract(p PrecompiledContract, input []byte, contract *Contract) ([]byte, uint64, error) {
-	gas, computationCost := p.GetRequiredGasAndComputationCost(input)
-	if contract.UseGas(gas) {
-		return contract.FeePayerAddress.Bytes(), computationCost, nil
-	}
-	return nil, computationCost, kerrors.ErrOutOfGas
+func (c *feePayer) Run(input []byte, contract *Contract, evm *EVM) ([]byte, error) {
+	return contract.FeePayerAddress.Bytes(), nil
 }
 
-type precompiledValidateSender struct{}
+type validateSender struct{}
 
-func (c *precompiledValidateSender) GetRequiredGasAndComputationCost(input []byte) (uint64, uint64) {
+func (c *validateSender) GetRequiredGasAndComputationCost(input []byte) (uint64, uint64) {
 	numSigs := uint64(len(input) / common.SignatureLength)
 	return numSigs * params.ValidateSenderGas,
 		numSigs*params.ValidateSenderPerSigComputationCost + params.ValidateSenderBaseComputationCost
 }
 
-func (c *precompiledValidateSender) Run(input []byte) ([]byte, error) {
-	// Run function should not be called. Instead of this function, RunValidateSenderContract should be called.
-	logger.Error("should not be reached here")
-	return nil, nil
-}
-
-func RunValidateSenderContract(p PrecompiledContract, input []byte, contract *Contract, picker types.AccountKeyPicker) ([]byte, uint64, error) {
-	// input must be (addr, message, signature1, signature2, and more signatures)
-	gas, computationCost := p.GetRequiredGasAndComputationCost(input)
-	if contract.UseGas(gas) == false {
-		return []byte{0}, computationCost, kerrors.ErrOutOfGas
-	}
-	if err := validateSender(input, picker); err != nil {
+func (c *validateSender) Run(input []byte, contract *Contract, evm *EVM) ([]byte, error) {
+	if err := validate(input, evm.StateDB); err != nil {
 		// If return error makes contract execution failed, do not return the error.
 		// Instead, print log.
 		logger.Trace("validateSender failed", "err", err)
-		return []byte{0}, computationCost, nil
+		return []byte{0}, nil
 	}
-
-	return []byte{1}, computationCost, nil
+	return []byte{1}, nil
 }
 
-func validateSender(input []byte, picker types.AccountKeyPicker) error {
+func validate(input []byte, picker types.AccountKeyPicker) error {
```

Comment:
Can I know why did you change this function name? I think 'validate' is too general and hard to find the connection with `validateSender` struct. How about using the original name or converting this function to a method of `validateSender` like `(c *validateSender) do`.

Reply:
I reflected your review at a96ffc7612723b72b3b3c9df756bb7bc3ddcfc27! Thank you.